File: take-gifts-from-the-richest-pile.ts
{
    function pickGifts(gifts: number[], k: number): number {
        while (k) {
            let max_index = 0;

            for (let i = 1; i < gifts.length; i++) 
                if (gifts[i] > gifts[max_index])
                    max_index = i;


            gifts[max_index] = ~~Math.sqrt(gifts[max_index]);
            k--;
        }

        return gifts.reduce((a, b) => a + b, 0);
    };

    console.log(pickGifts([25,64,9,4,100], 4))
}

File: take-gifts-from-the-richest-pile.rs
pub fn pick_gifts(mut gifts: Vec<i32>, mut k: i32) -> i64 {
    while k != 0 {
        let mut max_index = 0;

        for i in 1..gifts.len() {
            if gifts[i] > gifts[max_index] {
                max_index = i;
            }
        }

        let new_value = (gifts[max_index] as f64).sqrt().floor() as i32;
        
        gifts[max_index] = new_value;

        k -= 1;
    }

    gifts.into_iter().map(|x| x as i64).sum()
}